hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -225,13 +227,51 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
 	}
 
 	// Reconcile AWSMachineTemplate
-	// TODO (alberto): If a change happens to the nodePool AWSNodePoolPlatform we should
-	// delete the existing awsMachineTemplate, create a new one with a new name
+	// If a change happens to the nodePool AWSNodePoolPlatform we delete the existing awsMachineTemplate,
+	// create a new one with a new name
 	// and pass it to the machineDeployment. This will trigger a rolling upgrade.
-	if _, err := controllerutil.CreateOrPatch(ctx, r.Client, awsMachineTemplate, func() error {
-		return nil
-	}); err != nil {
-		return ctrl.Result{}, err
+	currentMD := &capiv1.MachineDeployment{}
+	if err := r.Get(ctx, ctrlclient.ObjectKeyFromObject(md), currentMD); err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, fmt.Errorf("failed to get machineDeployment: %w", err)
+	}
+
+	// If the machineDeployment has not been created yet, create new awsMachineTemplate.
+	if currentMD.CreationTimestamp.IsZero() {
+		r.Log.Info("Creating new AWSMachineTemplate", "AWSMachineTemplate", ctrlclient.ObjectKeyFromObject(awsMachineTemplate).String())
+		if _, err := controllerutil.CreateOrPatch(ctx, r.Client, awsMachineTemplate, func() error {
+			return nil
+		}); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error creating new AWSMachineTemplate: %w", err)
+		}
+	}
+
+	if !currentMD.CreationTimestamp.IsZero() {
+		currentAWSMachineTemplate := &capiaws.AWSMachineTemplate{}
+		if err := r.Get(ctx, client.ObjectKey{
+			Namespace: currentMD.Spec.Template.Spec.InfrastructureRef.Namespace,
+			Name:      currentMD.Spec.Template.Spec.InfrastructureRef.Name,
+		}, currentAWSMachineTemplate); err != nil && !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		if !equality.Semantic.DeepEqual(currentAWSMachineTemplate.Spec.Template.Spec, awsMachineTemplate.Spec.Template.Spec) {
+			r.Log.Info("AWS config has changed. This will trigger a rolling upgrade")
+			r.Log.Info("Creating new AWSMachineTemplate", "AWSMachineTemplate", ctrlclient.ObjectKeyFromObject(awsMachineTemplate).String())
+			// Create new template
+			if _, err := controllerutil.CreateOrPatch(ctx, r.Client, awsMachineTemplate, func() error {
+				return nil
+			}); err != nil {
+				return ctrl.Result{}, fmt.Errorf("error creating new AWSMachineTemplate: %w", err)
+			}
+			// Delete existing template
+			r.Log.Info("Deleting existing AWSMachineTemplate", "AWSMachineTemplate", ctrlclient.ObjectKeyFromObject(currentAWSMachineTemplate).String())
+			if err := r.Delete(ctx, currentAWSMachineTemplate); err != nil && !apierrors.IsNotFound(err) {
```

Comment:
should the delete happen when the MachineDeployment rollout has been completed?

Reply:
The machineDeployment should be able to react and go towards the new template anytime.

Reply:
What I mean is that while the machine deployment rolls out, a number of machines will reference a non-existent template. It may be ok for 3 machines, but if you have ~100 wouldn’t that be a problem?

Reply:
mmm good point, I wouldn't expect it to be a problem though, the `awsMachineTemplate` is only used by the underlying machineSet. When it instantiates new `Machines`, it also creates `awsMachines` out of the template.

Reply:
Thinking now this would be an issue for rollback situations though.

Reply:
It's fine if we want to leave it as a todo to address in a follow-up